### ISSUE TITLE
Suggestions for the UI install Document

### ISSUE THREAD

**Issue by stuclem** [vmware/vic-product#180]
@dbarkelew commented on [Wed Apr 26 2017](https://github.com/vmware/vic/issues/4898)

After a few UI plugin installs I have some feedback about the install docs. 

Feedback is for Document Links:
https://vmware.github.io/vic-product/assets/files/html/1.1/vic_vsphere_admin/plugins_vcsa.html
https://vmware.github.io/vic-product/assets/files/html/1.1/vic_vsphere_admin/plugins_vc_windows.html

The install requirements are not clear saying to perform install steps on the system the install was downloaded too.  This could be anything which is problematic. For example the install.sh cannot be run from ESXi due to not having curl to pass the check_prerequisite (), but based on how the docs are written it should work. Looking around I did not find any details of what OS can run the installer on other pages or where to run the installer from. Looking at each install.sh and install.bat, these look like they are easiest run on the vCenters directly and not from an external machine. Also I am seeing what looks like compatibility issues installing the UI to the VCSA from a windows box or installing the UI to a windows vCenter from a linux or Darwin OS. I hope to get a chance to test these later. 
Even so probably easier have the instructions linear showing a dependable example and as copy paste friendly as reasonable. For the sake of clarity draw the line in the sand of install from the vCenter being the go to as that is a fairly dependable configuration and installed set of packages. 

Windows vCenter:
First step is confusing both suggesting non vCenter system and using incorrect windows file syntax. 
Current - "1. On the Windows system on which you have downloaded and unpacked vSphere Integrated Containers Engine, open the \vic\ui\vCenterForWindows\configs file in a text editor."
Suggestion 
"1. On the vCenter Server open a browser to https://vic_appliance_address:9443. Download the vic 1.1.0.tar.zip and unpack the files to the desktop. Open the file %USERPROFILE%\Desktop\vic\ui\vCenterForWindows\configs in notepad."

**C:\WINDOWS\system32> notepad %USERPROFILE%\Desktop\vic\ui\vCenterForWindows\configs**

Suggesting to include a CLI way to get thumbprint for step 4. Then the current step 4 becomes step 5.

**C:\WINDOWS\system32> echo | "%VMWARE_OPENSSL_BIN%" s_client -connect vic_appliance_address:9443 | "%VMWARE_OPENSSL_BIN%" x509 -fingerprint -sha1 -noout**

In step 6 breaks syntax of previous examples. Would be more clear to use the full example of "SET target_vc_version=6.5" and "SET target_vc_version=6.0" respectively and keep a cleaner look over "6.5" and "6.0.

For copy/paste ability step 7. would update to something like "Open the Windows command prompt as administrator and run the install.sh. installer" with the follow for the sample.

**C:\WINDOWS\system32> %USERPROFILE%\Desktop\vic\ui\vCenterForWindows\install.bat**

In step 8. could help jog the end user's memory to include the example default user "administrator@vsphere.local" . Many users have a lot of administrator accounts for different parts of multiple VMware product offerings that things can get confusing. 

In step 9. suggest to also include step to restart web-client service as well if logging out and logging in does not activate the plugin. Often I find a newly registered plugin doesn't load as it should without starting up with the web client services start up processes.

**C:\WINDOWS\system32> "C:\Program Files\VMware\vCenter Server\bin\service-control" --stop vspherewebclientsvc
C:\WINDOWS\system32> "C:\Program Files\VMware\vCenter Server\bin\service-control" --start vspherewebclientsvc**

Last on the windows instructions, looking a the HTML5 there does not seem to be a compatible HTML5 installer the only one is bash. I do not have a windows 6.5 lab to test yet and would open separate issue if this turns out to be a problem.

VCSA:
Suggest including links the vCenter Bash Shell directions in the "Prerequisites" to the vSphere documents might help here in case this changes again. For example 6.0 originally had to enable Bash from DCUI, the vami was added in a later update (U1 I think).
6.0
http://pubs.vmware.com/vsphere-60/topic/com.vmware.vsphere.vcsa.doc/GUID-B259B969-3E5D-4576-8256-D3BED54B89AF.html
6.5
http://pubs.vmware.com/vsphere-65/topic/com.vmware.vsphere.vcsa.doc/GUID-B259B969-3E5D-4576-8256-D3BED54B89AF.html

Similar concerns for the first step as windows installer first step in having vague direction. Suggestion is to change the wording to be specific to the VCSA. For example.

1. SSH to the VCSA and download the vic ui installer and unpack the installer. 

**~# curl -k -o https://https://vic_appliance_address:9443/vic_1.1.0.tar.gz
~# tar -zxf vic_1.1.0.tar.gz** Name needs validation not 100% on the "_". I suppose the user could just download all three which wouldn't hurt. 

For each of the variables could also have a "sed" example as it is a bit more copy paste friendly for users who are not comfortable with VI due to the VCSA being the only linux VM like some Windows shops have. In this example all three config file locations can be updated from any previous value to the desired value in one command each.

**~ # sed -i 's#^\(VCENTER_IP=\).*$#\1"10.0.1.2"#' ~/vic/ui/*/configs
~ # sed -i 's#^\(VIC_UI_HOST_URL=\).*$#\1"https://vic_appliance_address:9443"#' ~/vic/ui/*/configs**

Recommend adding the CLI way to get thumbprint for step 4. The current step 4 becomes step 5. Although similar to the windows version '|&' is incompatible with Windows CLI.  

**~# echo | openssl s_client -connect vic_appliance_address:9443 |& openssl x509 -fingerprint -sha1 -noout**

Step 6 is confusing in the VCSA installer instruction. Suggest removing this step. If I am not mistaken even with the value change in vic/ui/HTML5Client/install.sh command prompt mentioned in step 7. will fail on the check due to having curl on windows. PowerShell has a curl alias, but my research is not showing bash to work on Windows in general. My initial testing with command prompt and PowerShell both fail to handle the install.sh.

Step 7 reference command prompt which is confusing seeing how this would be linux based instructions for most users. Maybe update to something like: 
"Run the install.sh installer"
_Flex Plugin_
**~# ~/vic/ui/VCSA/install.sh**
_HTML5 Plugin_
**~# ~/vic/ui/HTML5Client/install.sh**

Lastly would it be useful to have a note mentioning to check the performance of the web client is handling the additional load from plugin correctly? I have seen performance be factor for a plugin working or not. It is semi-common to see under sized vCenters and every so often the additional plugin load is the tipping point to start hitting data service timeouts. Section of possible interest "vSphere Web Clients" page 60 https://www.vmware.com/content/dam/digitalmarketing/vmware/en/pdf/techpaper/vmware-perfbest-practices-vsphere6-0-white-paper.pdf The one for 6.5 does not seem to have been release yet. 

---

@dbarkelew commented on [Wed Apr 26 2017](https://github.com/vmware/vic/issues/4898#issuecomment-297245934)

@stuclem You might be interested in the above feedback. 

---

@stuclem commented on [Wed Apr 26 2017](https://github.com/vmware/vic/issues/4898#issuecomment-297286186)

@dbarkelew I certainly am. Thank you for the very detailed feedback. I'm moving this issue to the `vic-product` repo, which is where we track doc updates.

**Comment by stuclem** [comment]
@dbarkelew, here are some initial responses to your questions/observations:

-  As far as I am aware, the installation of the UI is for vCenter Server deployments only. You can't deploy the UIs to an ESXi host because there is no Web Client for ESXi. @jooskim can you please confirm this? If so, I will make it clearer in the doc that ESXi does not support the client plugins.
- I agree that running the scripts on the machines where VC is running is probably easier, but this is not always the case, which is why I have included all the instances of "on the system on which you downloaded VIC engine" in the instructions, to make it clear that the system on which you run the script is not necessarily the same as the one where you run VC. My assumption was that users would run the scripts on the same system as they use to run `vic-machine`.
- As for compatibility issues when installing across platforms, this is how the scripts were written. Basically, to install the H5 plug-in on VC on Windows, you have no choice but to run `install.bat` on a Windows machine (e.g. the one where VC is running). Similarly, there is no `install.bat` script to install the H5 plug-in on a VCSA from a Windows machine. You either have to run `install.sh` from a Linux VM, log into the VCSA and run the script directly there as you suggest, or hack the `install.bat` script as described in the doc. I opted to document the option to hack the script, because that is how I got it to work when deploying from Windows to a VCSA. https://github.com/vmware/vic/issues/4692 is open in the `vic` repo about this.

**Comment by stuclem** [comment]
On the subject of the UI install, @pescerosso requests that I clarify the port number from which to download the package/cert.

**Comment by stuclem** [comment]
From @pdaigle via Slack:

```
Patrick Daigle [7:04 PM] 
I am having an issue installing the FLEX plugin with 1.1 (HTML5 works fine). figured I would check here before filing an issue.

[7:06] 
root@photon-b82rE9sWY [ ~/vic/ui/VCSA ]# ./install.sh
Enter your vCenter Administrator Username: administrator@corp.local
Enter your vCenter Administrator Password:
failed to connect: Failed to log in to vcsa-01a.corp.local: ServerFaultCode: Cannot complete login due to an incorrect user name or password.
Reading plugin-package.xml...
./install.sh: line 125: ../vsphere-client-serenity/-/vc_extension_flags: No such file or directory
File -.zip does not exist!

[7:06] 
I am using the same config I used for HTML5:

[7:07] 
# Enter the IP of the server running VCSA
VCENTER_IP=“vcsa-01a.corp.local”

# Update the following lines if you have a local web server that is accessible from your VCSA and VIC UI zip files are uploaded to it

# Change the following line to the full URL of the path where the zip files are located
# For example, if vic-ui.zip is available at https://192.168.1.5/vic-ui-plugins/vic-ui.zip, type in: https://192.168.1.5/vic-ui-plugins/
# Note: each uploaded zip file will be accessed only once when the user logs into vSphere Web Client for the first time after successful installation
VIC_UI_HOST_URL=“https://192.168.100.102:9443"

Josh Kim
[7:16 PM] 
@pdaigle Hi Patrick, I admit I forgot to update the installer for Flex plugin as I previously made some changes to the H5 plugin installer to make it usable for the both plugins. So you could use the H5 plugin installer to install the Flex plugin. Open up the `HTML5Client/configs` file and add `plugin_type=flex` at the end, then run `HTML5Client/install.sh`.

[7:17] 
@sclements we need to update the docs for client plugin as they currently mention to start the install.sh script in VSCSA directory
```

**Comment by dbarkelew** [comment]
I am working on a similar issue with "./install.sh: line 125: ../vsphere-client-serenity/-/vc_extension_flags: No such file or directory
File -.zip does not exist!". 

I am looking to this today and hoping to have an answer by today or tomorrow.

**Comment by jooskim** [comment]
@stuclem Yes, the UI plugin is supposed to be installed and deployed on a VCSA or vCenter Server for Windows but not an ESXi instance. Regarding your third point please see the latest comment I made.

**Comment by dbarkelew** [comment]
Seems a backslash was dropped from the sed commands in the copy paste unless looking at the edit of the original post. Seems |& is not working with VCSA although |& works on other linux distros .

**~ # sed -i 's#^(VCENTER_IP=\).$#\1"10.0.1.2"#' ~/vic/ui//configs**
**~ # sed -i 's#^(VIC_UI_HOST_URL=\).$#\1"https://vic_appliance_address:9443"#' ~/vic/ui//configs**
**~# echo | openssl s_client -connect vic_appliance_address:9443 |& openssl x509 -fingerprint -sha1 -noout**

Correct commands in code blocks below and updated the original to code blocks:
 
`~# echo | openssl s_client -connect vic_appliance_address:9443 | openssl x509 -fingerprint -sha1 -noout`

`sed -i 's#^\(VCENTER_IP=\).*$#\1"10.0.1.2"#' ~/vic/ui/*/configs`
`sed -i 's#^\(VIC_UI_HOST_URL=\).*$#\1"https://vic_appliance_address:9443"#' ~/vic/ui/*/configs`

**Comment by jooskim** [comment]
I just finished reading @dbarkelew 's detailed feedback. Here is a change I merged onto the master branch yesterday regarding this issue:

1. `HTML5Client` folder is now removed and consolidated into the `VCSA` folder
2. `VCSA/configs` is also updated to include more explicit instructions and cautions
3. `VCSA/configs` now includes the PLUGIN_TYPE variable with a default value of "html5". This could be changed to "flex" to allow for flex plugin installation

I agree with Dani's aforementioned suggestions and that we need to have linear samples the for most common cases. Dani could you confirm the following?

1. Most VCSA admins will (and prefer to) SSH into the VCSA host to install the VIC
2. Most vCenter for Windows admins will (and prefer to) RDP to the vCenter host to install VIC

**Comment by stuclem** [comment]
Thanks @jooskim, I will update the docs accordingly in time for 1.1.1.

**Comment by jooskim** [comment]
FYI https://github.com/vmware/vic/pull/4947

**Comment by stuclem** [comment]
See also https://github.com/vmware/vic/issues/4948

**Comment by stuclem** [comment]
From @dbarkelew by email:

First the VCSA install.sh is broken. Work around is to use the HTML5 installer.sh change the ui/installer/HTML5Client/ to append plugin_type="flex" (lowercase for this one unlike the other values). I will be using this with a customer in a few minutes from now. 

Other concerns is how the documentation is presented. We discussed having installing on the vCenter being the goto example for both windows and VCSA types. Then later having supplemental information about the remote install option for running the installer from non-vcenter machines. Could be in a section dedicated to automated install/deployments or blog post or KB or just an advanced section. 

At the immediate the feedback I have from customers is the docs are not “clear”. In translation what they really mean is the documents leave too many choices for them to figure out. The docs have all the info (bug aside) but require a higher than the average foundation knowledge and datacenter experience to fill in the choices that need to be made. 

Josh is taking some time to consider having vCenter as the go to place to run the ui installs and keep the rest of the abilities in an advanced section should we ever be presented with a use case for this aside from internal automated testing. 

The more the instructions can be broken down to essentially copy paste level the better.

**Comment by vmwarelab** [comment]
curl -k -o https://https://vic_appliance_address:9443/vic_1.1.0.tar.gz

this command doenst seem right .. duplicate https:// i tried both and i get an error 
curl: no URL specified!
curl: try 'curl --help' or 'curl --manual' for more information

**Comment by dbarkelew** [comment]
The 'o' need to be capital sorry. 

`curl -k -O https://vic_appliance_address:9443/vic_1.1.0.tar.gz`

**Comment by stuclem** [comment]
See also https://github.com/vmware/vic/issues/4981
